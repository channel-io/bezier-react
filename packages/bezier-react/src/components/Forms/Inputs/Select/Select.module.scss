.SelectContainer {
  position: relative;
}

.SelectTrigger {
  all: unset;

  cursor: pointer;
  user-select: none;

  display: flex;
  align-items: center;
  justify-content: space-between;

  box-sizing: border-box;
  width: 100%;
  padding: 8px 12px;

  color: var(--txt-black-darkest);

  background-color: var(--bg-grey-lightest);
  border-radius: var(--radius-8);
  box-shadow: var(--input-box-shadow);

  transition: background-color var(--transition-s), box-shadow var(--transition-s);

  &:where(.size-28) {
    height: 28px;
  }

  &:where(.size-36) {
    height: 36px;
  }

  &:where(.size-44) {
    height: 44px;
  }

  &:where(.size-54) {
    height: 54px;
  }

  &.readonly {
    cursor: initial;
    color: var(--txt-black-darker);
    background-color: var(--bg-grey-lighter);
  }

  &.invalid {
    box-shadow: var(--input-box-shadow-invalid);
  }

  &:where(:disabled) {
    cursor: not-allowed;
    opacity: var(--opacity-disabled);
  }

  &:where(:not(:disabled):where(.active, :focus)) {
    background-color: var(--bg-grey-lighter);
    box-shadow: var(--input-box-shadow-focused);
  }

  &:where(:not(.readonly, :disabled):hover) {
    background-color: var(--bg-grey-lighter);
  }
}

.SelectMainContent {
  overflow: hidden;
  display: flex;
  align-items: center;
}

.SelectDropdown {
  min-width: 200px;
  min-height: 42px;
  max-height: 640px;

  background-color: var(--bg-white-high);
  border-radius: var(--radius-8);
  box-shadow: var(--ev-3);
}
