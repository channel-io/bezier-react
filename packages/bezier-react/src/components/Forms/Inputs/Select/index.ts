export { Select } from './Select'

export type {
  SelectProps,
  SelectRef,
} from './Select.types'
