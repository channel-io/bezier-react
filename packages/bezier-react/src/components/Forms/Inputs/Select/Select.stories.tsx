import React from 'react'

import { CalendarIcon } from '@channel.io/bezier-icons'
import {
  type Meta,
  type StoryFn,
} from '@storybook/react'

import { getObjectFromEnum } from '~/src/utils/story'

import { Text } from '~/src/components/Text'

import { Select } from './Select'
import {
  type SelectProps,
  SelectSize,
} from './Select.types'

const meta: Meta<SelectProps & {
  wrapperSize: number
}> = {
  component: Select,
  argTypes: {
    size: {
      control: {
        type: 'radio',
      },
      options: getObjectFromEnum(SelectSize),
    },
    wrapperSize: {
      control: {
        type: 'number',
      },
      defaultValue: 200,
    },
  },
}
export default meta

const Template: StoryFn<SelectProps & { wrapperSize: number }> = ({
  wrapperSize,
  ...args
}) => (
  <div style={{ width: wrapperSize }}>
    <Select {...args} />
  </div>
)

export const Primary = {
  render: Template,

  args: {
    placeholder: '날짜를 선택해주세요',
    text: '2022. 7. 14.',
    leftContent: CalendarIcon,
    rightContent: <Text marginLeft={4}>일</Text>,
    disabled: false,
    readOnly: false,
    withoutChevron: false,
    hasError: false,
    size: SelectSize.M,
  },
}
