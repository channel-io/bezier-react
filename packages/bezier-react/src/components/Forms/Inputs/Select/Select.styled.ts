import {
  css,
  styled,
} from '~/src/foundation'

import type { InterpolationProps } from '~/src/types/Foundation'

import {
  erroredInputWrapperStyle,
  focusedInputWrapperStyle,
  inputTextStyle,
  inputWrapperStyle,
} from '~/src/components/Forms/Inputs/mixins'
import { Overlay } from '~/src/components/Overlay'

import { type SelectSize } from './Select.types'

export const Container = styled.div<InterpolationProps>`
  position: relative;

  ${({ interpolation }) => interpolation};
`

interface TriggerProps {
  hasError: boolean
  disabled: boolean
  readOnly: boolean
  active: boolean
  size: SelectSize
}

const focusedStyle = css`
  ${focusedInputWrapperStyle};
  background-color: ${({ foundation }) => foundation?.theme?.['bg-grey-lighter']};
`

export const Trigger = styled.button<TriggerProps>`
  all: unset;
  box-sizing: border-box;
  display: flex;
  align-items: center;
  justify-content: space-between;
  width: 100%;
  height: ${({ size }) => size}px;
  padding: 8px 12px;
  cursor: pointer;
  user-select: none;
  background-color: ${({ foundation }) => foundation?.theme?.['bg-grey-lightest']};

  ${inputTextStyle}

  ${inputWrapperStyle}

  ${({ foundation }) => foundation?.rounding?.round8}

  ${({ foundation }) => foundation?.transition?.getTransitionsCSS(['background-color', 'box-shadow'])}

  ${({ active }) => active && focusedStyle}

  ${({ hasError }) => hasError && erroredInputWrapperStyle}

  ${({ readOnly }) => readOnly && css`
    cursor: initial;
    background-color: ${({ foundation }) => foundation?.theme?.['bg-grey-lighter']};
  `}

  &:disabled {
    cursor: not-allowed;
    opacity: var(--opacity-disabled);
  }

  &:not(:disabled):focus {
    ${focusedStyle}
  }

  ${({ readOnly }) => !readOnly && css`
    &:not(:disabled):hover {
      background-color: ${({ foundation }) => foundation?.theme?.['bg-grey-lighter']};
    }
  `}
`

export const MainContentWrapper = styled.div`
  display: flex;
  align-items: center;
  overflow: hidden;
`

interface DropdownProps extends InterpolationProps {
  zIndex: number
}

export const Dropdown = styled(Overlay)<DropdownProps>`
  z-index: ${({ zIndex }) => zIndex};
  min-width: 200px;
  min-height: 42px;
  max-height: 640px;

  ${({ foundation }) => foundation?.rounding?.round8};
  ${({ foundation }) => foundation?.elevation?.ev3()};

  ${({ interpolation }) => interpolation}
`
