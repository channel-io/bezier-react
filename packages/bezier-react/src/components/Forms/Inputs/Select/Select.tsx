/* External dependencies */
import React, {
  useState,
  useCallback,
  useMemo,
  useRef,
  forwardRef,
  Ref,
  useImperativeHandle,
  useEffect,
} from 'react'

/* Internal dependencies */
import { Typography } from '~/src/foundation'
import { ZIndex } from '~/src/constants/ZIndex'
import {
  isEmpty,
} from '~/src/utils/typeUtils'
import { noop } from '~/src/utils/functionUtils'
import {
  Icon,
  ChevronUpIcon,
  ChevronDownIcon,
  IconSize,
  isBezierIcon,
  isIconName,
  LegacyIcon,
} from '~/src/components/Icon'
import { OverlayPosition } from '~/src/components/Overlay'
import { Text } from '~/src/components/Text'
import useFormFieldProps from '~/src/components/Forms/useFormFieldProps'

import SelectProps, { SelectRef, SelectSize } from './Select.types'
import * as Styled from './Select.styled'

export const SELECT_CONTAINER_TEST_ID = 'bezier-react-select-container'
export const SELECT_TRIGGER_TEST_ID = 'bezier-react-select-trigger'
export const SELECT_TRIGGER_TEXT_TEST_ID = 'bezier-react-select-trigger-text'
export const SELECT_DROPDOWN_TEST_ID = 'bezier-react-select-dropdown'

const DEFAULT_DROPDOWN_MARGIN_Y = 6
const DEFAULT_DROPDOWN_Z_INDEX = ZIndex.Overlay

function Select({
  testId = SELECT_CONTAINER_TEST_ID,
  triggerTestId = SELECT_TRIGGER_TEST_ID,
  triggerTextTestId = SELECT_TRIGGER_TEXT_TEST_ID,
  dropdownTestId = SELECT_DROPDOWN_TEST_ID,
  as,
  style,
  className,
  interpolation,
  dropdownInterpolation,
  size = SelectSize.M,
  defaultFocus = false,
  placeholder = '',
  leftContent,
  rightContent,
  iconColor = 'txt-black-dark',
  text,
  textColor = 'txt-black-darkest',
  withoutChevron = false,
  dropdownStyle,
  dropdownClassName,
  dropdownContainer,
  dropdownMarginX,
  dropdownMarginY = DEFAULT_DROPDOWN_MARGIN_Y,
  dropdownZIndex = DEFAULT_DROPDOWN_Z_INDEX,
  dropdownPosition = OverlayPosition.BottomLeft,
  onClickTrigger = noop,
  onHideDropdown = noop,
  children,
  ...rest
}: SelectProps,
forwardedRef: Ref<SelectRef>,
) {
  const {
    disabled,
    readOnly,
    hasError,
    ...ownProps
  } = useFormFieldProps(rest)

  const containerRef = useRef<HTMLDivElement>(null)
  const triggerRef = useRef<HTMLDivElement>(null)

  const [isDropdownOpened, setIsDropdownOpened] = useState(false)

  const LeftComponent = useMemo(() => {
    if (isBezierIcon(leftContent)) {
      return (
        <Icon
          source={leftContent}
          size={IconSize.XS}
          marginRight={6}
          color={iconColor}
        />
      )
    }

    if (isIconName(leftContent)) {
      return (
        <LegacyIcon
          name={leftContent}
          size={IconSize.XS}
          color={iconColor}
          marginRight={6}
        />
      )
    }

    return leftContent
  }, [
    leftContent,
    iconColor,
  ])

  const RightComponent = useMemo(() => {
    if (isBezierIcon(rightContent)) {
      return (
        <Icon
          source={rightContent}
          size={IconSize.XS}
          marginRight={6}
          color={iconColor}
        />
      )
    }

    if (isIconName(rightContent)) {
      return (
        <LegacyIcon
          name={rightContent}
          size={IconSize.XS}
          color={iconColor}
          marginRight={6}
        />
      )
    }

    return rightContent
  }, [
    rightContent,
    iconColor,
  ])

  const handleClickTrigger = useCallback((event: React.MouseEvent) => {
    if (!disabled && !readOnly) {
      setIsDropdownOpened(prevState => !prevState)
      onClickTrigger(event)
    }
  }, [
    disabled,
    readOnly,
    onClickTrigger,
  ])

  const handleHideDropdown = useCallback(() => {
    setIsDropdownOpened(false)
    onHideDropdown()
  }, [onHideDropdown])

  const getDOMNode = useCallback(() => triggerRef.current, [])

  const handle = useMemo((): SelectRef => ({ handleClickTrigger, handleHideDropdown, getDOMNode }), [
    handleClickTrigger,
    handleHideDropdown,
    getDOMNode,
  ])

  useEffect(() => {
    if (defaultFocus && !disabled && !readOnly) {
      setIsDropdownOpened(true)
    }
  }, [
    defaultFocus,
    disabled,
    readOnly,
  ])

  useImperativeHandle(forwardedRef, () => handle)

  const hasContent = !isEmpty(text)

  return (
    <Styled.Container
      data-testid={testId}
      ref={containerRef}
      style={style}
      className={className}
      interpolation={interpolation}
    >
      <Styled.Trigger
        type="button"
        data-testid={triggerTestId}
        as={as}
        ref={triggerRef}
        size={size}
        hasError={hasError}
        disabled={disabled}
        readOnly={readOnly}
        active={isDropdownOpened}
        onClick={handleClickTrigger}
        {...ownProps}
      >
        <Styled.MainContentWrapper>
          { LeftComponent }
          <Text
            testId={triggerTextTestId}
            typo={Typography.Size14}
            truncated
            color={hasContent ? textColor : 'txt-black-dark'}
          >
            { hasContent ? text : placeholder }
          </Text>
          { RightComponent }
        </Styled.MainContentWrapper>
        { !withoutChevron && (
          <Icon
            source={isDropdownOpened ? ChevronUpIcon : ChevronDownIcon}
            size={IconSize.XS}
            color={readOnly ? 'txt-black-dark' : 'txt-black-darker'}
            marginLeft={6}
          />
        ) }
      </Styled.Trigger>

      <Styled.Dropdown
        style={dropdownStyle}
        className={dropdownClassName}
        testId={dropdownTestId}
        withTransition
        show={isDropdownOpened && !disabled}
        zIndex={dropdownZIndex}
        marginX={dropdownMarginX}
        marginY={dropdownMarginY}
        target={triggerRef.current}
        container={dropdownContainer || containerRef.current}
        position={dropdownPosition}
        interpolation={dropdownInterpolation}
        onHide={handleHideDropdown}
      >
        { children }
      </Styled.Dropdown>
    </Styled.Container>
  )
}

export default forwardRef(Select)
