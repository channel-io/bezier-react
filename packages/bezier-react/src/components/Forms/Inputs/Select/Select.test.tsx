/* External dependencies */
import React from 'react'
import { fireEvent } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

/* Internal dependencies */
import { LightFoundation } from 'Foundation'
import { render } from 'Utils/testUtils'
import { ZIndex } from 'Constants/ZIndex'
import Select, {
  SELECT_CONTAINER_TEST_ID,
  SELECT_TRIGGER_TEST_ID,
  SELECT_DROPDOWN_TEST_ID,
  SELECT_TRIGGER_TEXT_TEST_ID,
} from './Select'
import SelectProps, { SelectSize } from './Select.types'

describe('Select Test >', () => {
  let props: SelectProps

  beforeEach(() => {
    const { body } = document
    body.innerHTML = ''

    const rootElement = document.createElement('div')
    rootElement.setAttribute('id', 'root')

    body.appendChild(rootElement)
  })

  const renderSelect = (optionProps?: Partial<SelectProps>) => render(<Select {...props} {...optionProps} />)

  describe('Event >', () => {
    const renderFormWithSelect = (onSubmit: React.FormEventHandler) => render(
      <form onSubmit={onSubmit}>
        <Select {...props} />
      </form>,
    )

    it('should fire click event when clicking the trigger button', () => {
      const onClick = jest.fn()
      const { getByTestId } = renderSelect({ onClickTrigger: onClick })
      const rendered = getByTestId(SELECT_TRIGGER_TEST_ID)

      fireEvent.click(rendered)
      expect(onClick).toHaveBeenCalled()
    })

    it('should not fire submit event when clicking the trigger button', () => {
      const onSubmit = jest.fn()
      const { getByTestId } = renderFormWithSelect(onSubmit)
      const rendered = getByTestId(SELECT_TRIGGER_TEST_ID)

      fireEvent.click(rendered)
      expect(onSubmit).not.toHaveBeenCalled()
    })
  })

  describe('Default Style >', () => {
    it('Snapshot >', () => {
      const { container } = renderSelect({ text: 'foo' })
      expect(container.firstChild).toMatchSnapshot()
    })

    it('Container >', () => {
      const { getByTestId } = renderSelect()
      const defaultSelect = getByTestId(SELECT_CONTAINER_TEST_ID)

      expect(defaultSelect).toHaveStyle('position: relative;')
    })

    it('Trigger >', () => {
      const { getByTestId } = renderSelect()
      const defaultSelectTrigger = getByTestId(SELECT_TRIGGER_TEST_ID)

      expect(defaultSelectTrigger).toHaveStyle('display: flex;')
      expect(defaultSelectTrigger).toHaveStyle('align-items: center;')
      expect(defaultSelectTrigger).toHaveStyle('width: 100%;')
      expect(defaultSelectTrigger).toHaveStyle('justify-content: space-between;')
      expect(defaultSelectTrigger).toHaveStyle('padding: 8px 12px;')
      expect(defaultSelectTrigger).toHaveStyle('cursor: pointer;')
    })

    it('Dropdown >', async () => {
      const user = userEvent.setup()
      const { getByTestId } = renderSelect()

      const defaultSelectTrigger = getByTestId(SELECT_TRIGGER_TEST_ID)
      await user.click(defaultSelectTrigger)

      const defaultSelectDropdown = getByTestId(SELECT_DROPDOWN_TEST_ID)

      expect(defaultSelectDropdown).toHaveStyle(`z-index: ${ZIndex.Overlay};`)
      expect(defaultSelectDropdown).toHaveStyle('min-width: 200px;')
      expect(defaultSelectDropdown).toHaveStyle('min-height: 42px;')
      expect(defaultSelectDropdown).toHaveStyle('max-height: 640px;')
    })
  })

  describe('SelectSize >', () => {
    it('Size only effects to trigger style >', async () => {
      const user = userEvent.setup()
      const XL_CONTAINER_TEST_ID = `${SELECT_CONTAINER_TEST_ID}_XL`
      const XL_TRIGGER_TEST_ID = `${SELECT_TRIGGER_TEST_ID}_XL`
      const XL_DROPDOWN_TEST_ID = `${SELECT_DROPDOWN_TEST_ID}_XL`
      const L_CONTAINER_TEST_ID = `${SELECT_CONTAINER_TEST_ID}_L`
      const L_TRIGGER_TEST_ID = `${SELECT_TRIGGER_TEST_ID}_L`
      const L_DROPDOWN_TEST_ID = `${SELECT_DROPDOWN_TEST_ID}_L`

      const { getByTestId: getByTestIdForXL } = renderSelect({
        size: SelectSize.XL,
        testId: XL_CONTAINER_TEST_ID,
        triggerTestId: XL_TRIGGER_TEST_ID,
        dropdownTestId: XL_DROPDOWN_TEST_ID,
      })
      const { getByTestId: getByTestIdForL } = renderSelect({
        size: SelectSize.L,
        testId: L_CONTAINER_TEST_ID,
        triggerTestId: L_TRIGGER_TEST_ID,
        dropdownTestId: L_DROPDOWN_TEST_ID,
      })

      // Select Size.XL
      const xlSelectContainer = getByTestIdForXL(XL_CONTAINER_TEST_ID)
      const xlSelectTrigger = getByTestIdForXL(XL_TRIGGER_TEST_ID)
      // Open Dropdown
      await user.click(xlSelectTrigger)
      const xlSelectDropdown = getByTestIdForXL(XL_DROPDOWN_TEST_ID)
      // Close Dropdown
      await user.click(xlSelectTrigger)
      const xlContainerStyle = window.getComputedStyle(xlSelectContainer)
      const xlDropdownStyle = window.getComputedStyle(xlSelectDropdown)

      // Select Size.L
      const lSelectContainer = getByTestIdForL(L_CONTAINER_TEST_ID)
      const lSelectTrigger = getByTestIdForL(L_TRIGGER_TEST_ID)
      // Open Dropdown
      await user.click(lSelectTrigger)
      const lSelectDropdown = getByTestIdForL(L_DROPDOWN_TEST_ID)
      // Close Dropdown
      await user.click(lSelectTrigger)
      const lContainerStyle = window.getComputedStyle(lSelectContainer)
      const lDropdownStyle = window.getComputedStyle(lSelectDropdown)

      // Container
      expect(xlContainerStyle.getPropertyValue('display')).toEqual(lContainerStyle.getPropertyValue('display'))

      // Dropdown
      expect(xlDropdownStyle.getPropertyValue('width')).toEqual(lDropdownStyle.getPropertyValue('width'))
      expect(xlDropdownStyle.getPropertyValue('min-width')).toEqual(lDropdownStyle.getPropertyValue('min-width'))
      expect(xlDropdownStyle.getPropertyValue('min-height')).toEqual(lDropdownStyle.getPropertyValue('min-height'))
      expect(xlDropdownStyle.getPropertyValue('max-height')).toEqual(lDropdownStyle.getPropertyValue('max-height'))
    })

    it('Size S >', () => {
      const { getByTestId } = renderSelect({ size: SelectSize.S })
      const defaultSelectDropdown = getByTestId(SELECT_TRIGGER_TEST_ID)

      expect(defaultSelectDropdown).toHaveStyle('height: 28px;')
    })

    it('Size M >', () => {
      const { getByTestId } = renderSelect({ size: SelectSize.M })
      const defaultSelectDropdown = getByTestId(SELECT_TRIGGER_TEST_ID)

      expect(defaultSelectDropdown).toHaveStyle('height: 36px;')
    })

    it('Size L >', () => {
      const { getByTestId } = renderSelect({ size: SelectSize.L })
      const defaultSelectDropdown = getByTestId(SELECT_TRIGGER_TEST_ID)

      expect(defaultSelectDropdown).toHaveStyle('height: 44px;')
    })

    it('Size XL >', () => {
      const { getByTestId } = renderSelect({ size: SelectSize.XL })
      const defaultSelectDropdown = getByTestId(SELECT_TRIGGER_TEST_ID)

      expect(defaultSelectDropdown).toHaveStyle('height: 54px;')
    })
  })

  describe('Trigger text style >', () => {
    it('shows default text color', () => {
      const { getByTestId } = renderSelect({ text: 'lorem ipsum' })
      const triggerText = getByTestId(SELECT_TRIGGER_TEXT_TEST_ID)

      expect(triggerText).toHaveStyle(`color: ${LightFoundation.theme['txt-black-darkest']};`)
    })

    it('shows given text color', () => {
      const { getByTestId } = renderSelect({ text: 'lorem ipsum', textColor: 'bgtxt-green-normal' })
      const triggerText = getByTestId(SELECT_TRIGGER_TEXT_TEST_ID)

      expect(triggerText).toHaveStyle(`color: ${LightFoundation.theme['bgtxt-green-normal']};`)
    })

    it('shows placeholder text color if content is empty', () => {
      const { getByTestId } = renderSelect({ text: '', textColor: 'bgtxt-blue-normal' })
      const triggerText = getByTestId(SELECT_TRIGGER_TEXT_TEST_ID)

      expect(triggerText).toHaveStyle(`color: ${LightFoundation.theme['txt-black-dark']};`)
    })
  })

  describe('Disabled style', () => {
    it('should have dimmed opacity', () => {
      const { getByTestId } = renderSelect({ disabled: true })
      const trigger = getByTestId(SELECT_TRIGGER_TEST_ID)

      expect(trigger).toHaveStyle('opacity: 0.4')
    })

    it('should have not-allowed cursor style', () => {
      const { getByTestId } = renderSelect({ disabled: true })
      const trigger = getByTestId(SELECT_TRIGGER_TEST_ID)

      expect(trigger).toHaveStyle('cursor: not-allowed')
    })

    it('should not show dropdown when clicked', async () => {
      const user = userEvent.setup()
      const { getByTestId } = renderSelect({ disabled: true })
      const trigger = getByTestId(SELECT_TRIGGER_TEST_ID)

      await user.click(trigger)

      expect(() => getByTestId(SELECT_DROPDOWN_TEST_ID)).toThrow() // element should not exist
    })

    it('should not update style when hovered', () => {
      const { getByTestId } = renderSelect({ disabled: true })
      const trigger = getByTestId(SELECT_TRIGGER_TEST_ID)

      fireEvent.mouseOver(trigger)

      expect(trigger).toHaveStyle(`background-color: ${LightFoundation.theme['bg-grey-lightest']}`)
    })

    it('overrides read-only style', () => {
      const { getByTestId } = renderSelect({ readOnly: true, disabled: true })
      const trigger = getByTestId(SELECT_TRIGGER_TEST_ID)

      expect(trigger).toHaveStyle('opacity: 0.4')
    })
  })

  describe('Read-only style', () => {
    it('should have regular opacity', () => {
      const { getByTestId } = renderSelect({ readOnly: true })
      const trigger = getByTestId(SELECT_TRIGGER_TEST_ID)

      expect(trigger).not.toHaveStyle('opacity: 0.4')
    })

    it('should have initial cursor style', () => {
      const { getByTestId } = renderSelect({ readOnly: true })
      const trigger = getByTestId(SELECT_TRIGGER_TEST_ID)

      expect(trigger).toHaveStyle('cursor: initial')
    })

    it('should not show dropdown when clicked', async () => {
      const user = userEvent.setup()
      const { getByTestId } = renderSelect({ readOnly: true })
      const trigger = getByTestId(SELECT_TRIGGER_TEST_ID)

      await user.click(trigger)

      expect(() => getByTestId(SELECT_DROPDOWN_TEST_ID)).toThrow() // element should not exist
    })

    it('should have chevron with txt-black-dark color', () => {
      const { getByTestId } = renderSelect({ readOnly: true })
      const trigger = getByTestId(SELECT_TRIGGER_TEST_ID)

      expect(trigger.children.item(1)).toHaveStyle(`color: ${LightFoundation.theme['txt-black-dark']}`)
    })
  })

  describe('defaultFocus >', () => {
    it('should not show dropdown when the first time If the defaultFocus is false >', () => {
      const { queryByTestId } = renderSelect({ defaultFocus: false })
      const dropdown = queryByTestId(SELECT_DROPDOWN_TEST_ID)

      expect(dropdown).toBeNull()
    })

    it('should show dropdown when the first time If the defaultFocus is true >', () => {
      const { getByTestId } = renderSelect({ defaultFocus: true })
      const dropdown = getByTestId(SELECT_DROPDOWN_TEST_ID)

      expect(dropdown).toBeVisible()
    })

    it('should disappear dropdown when the other elements clicked >', async () => {
      const { body } = document
      const user = userEvent.setup()

      const { getByTestId } = renderSelect({ defaultFocus: true })
      const dropdown = getByTestId(SELECT_DROPDOWN_TEST_ID)

      await user.click(body)

      expect(dropdown).not.toBeVisible()
    })
  })

  describe('rightContent >', () => {
    it('Snapshot >', () => {
      const { container } = renderSelect({ text: 'lorem ipsum', rightContent: (<div>가나다</div>) })
      expect(container.firstChild).toMatchSnapshot()
    })
  })
})
