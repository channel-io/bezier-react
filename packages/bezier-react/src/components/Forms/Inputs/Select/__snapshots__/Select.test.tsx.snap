// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select Test > Default Style > Snapshot > 1`] = `
.c5 {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  margin: 0px 0px 0px 6px;
  color: #00000099;
  -webkit-transition-delay: 0ms;
  transition-delay: 0ms;
  -webkit-transition-timing-function: cubic-bezier(.3,0,0,1);
  transition-timing-function: cubic-bezier(.3,0,0,1);
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  -webkit-transition-property: color;
  transition-property: color;
}

.c3 {
  font-size: 1.4rem;
  line-height: 1.8rem;
  margin: 0px 0px 0px 0px;
  font-style: normal;
  font-weight: normal;
  color: #000000D9;
  -webkit-transition-delay: 0ms;
  transition-delay: 0ms;
  -webkit-transition-timing-function: cubic-bezier(.3,0,0,1);
  transition-timing-function: cubic-bezier(.3,0,0,1);
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  -webkit-transition-property: color;
  transition-property: color;
}

.c0 {
  position: relative;
}

.c1 {
  all: unset;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  width: 100%;
  height: 36px;
  padding: 8px 12px;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  background-color: #FCFCFC;
  color: #000000D9;
  box-shadow: 0 1px 2px #0000000D, inset 0 0 0 1px #00000026;
  overflow: hidden;
  border-radius: 8px;
  -webkit-transition-delay: 0ms;
  transition-delay: 0ms;
  -webkit-transition-timing-function: cubic-bezier(.3,0,0,1);
  transition-timing-function: cubic-bezier(.3,0,0,1);
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  -webkit-transition-property: background-color,box-shadow;
  transition-property: background-color,box-shadow;
}

.c1:disabled {
  cursor: not-allowed;
  opacity: 0.4;
}

.c1:not(:disabled):focus {
  box-shadow: 0 0 0 3px #5E56F04D, inset 0 0 0 1px #5E56F0;
  background-color: #F7F7F8;
}

.c1:not(:disabled):hover {
  background-color: #F7F7F8;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  overflow: hidden;
}

.c4 {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

<div
  class="c0"
  data-testid="bezier-react-select-container"
>
  <button
    class="c1"
    data-testid="bezier-react-select-trigger"
    size="36"
    type="button"
  >
    <div
      class="c2"
    >
      <span
        class="c3 c4"
        color="txt-black-darkest"
        data-testid="bezier-react-select-trigger-text"
      >
        foo
      </span>
    </div>
    <svg
      class="c5"
      color="txt-black-darker"
      data-testid="bezier-react-icon"
      fill="none"
      foundation="[object Object]"
      height="16"
      marginbottom="0"
      marginleft="6"
      marginright="0"
      margintop="0"
      viewBox="0 0 24 24"
      width="16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clip-rule="evenodd"
        d="M19.707 8.793a1 1 0 0 1 0 1.414l-7 7a1 1 0 0 1-1.414 0l-7-7a1 1 0 0 1 1.414-1.414L12 15.086l6.293-6.293a1 1 0 0 1 1.414 0Z"
        fill="currentColor"
        fill-rule="evenodd"
      />
    </svg>
  </button>
</div>
`;

exports[`Select Test > rightContent > Snapshot > 1`] = `
.c5 {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  margin: 0px 0px 0px 6px;
  color: #00000099;
  -webkit-transition-delay: 0ms;
  transition-delay: 0ms;
  -webkit-transition-timing-function: cubic-bezier(.3,0,0,1);
  transition-timing-function: cubic-bezier(.3,0,0,1);
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  -webkit-transition-property: color;
  transition-property: color;
}

.c3 {
  font-size: 1.4rem;
  line-height: 1.8rem;
  margin: 0px 0px 0px 0px;
  font-style: normal;
  font-weight: normal;
  color: #000000D9;
  -webkit-transition-delay: 0ms;
  transition-delay: 0ms;
  -webkit-transition-timing-function: cubic-bezier(.3,0,0,1);
  transition-timing-function: cubic-bezier(.3,0,0,1);
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  -webkit-transition-property: color;
  transition-property: color;
}

.c0 {
  position: relative;
}

.c1 {
  all: unset;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  width: 100%;
  height: 36px;
  padding: 8px 12px;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  background-color: #FCFCFC;
  color: #000000D9;
  box-shadow: 0 1px 2px #0000000D, inset 0 0 0 1px #00000026;
  overflow: hidden;
  border-radius: 8px;
  -webkit-transition-delay: 0ms;
  transition-delay: 0ms;
  -webkit-transition-timing-function: cubic-bezier(.3,0,0,1);
  transition-timing-function: cubic-bezier(.3,0,0,1);
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  -webkit-transition-property: background-color,box-shadow;
  transition-property: background-color,box-shadow;
}

.c1:disabled {
  cursor: not-allowed;
  opacity: 0.4;
}

.c1:not(:disabled):focus {
  box-shadow: 0 0 0 3px #5E56F04D, inset 0 0 0 1px #5E56F0;
  background-color: #F7F7F8;
}

.c1:not(:disabled):hover {
  background-color: #F7F7F8;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  overflow: hidden;
}

.c4 {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

<div
  class="c0"
  data-testid="bezier-react-select-container"
>
  <button
    class="c1"
    data-testid="bezier-react-select-trigger"
    size="36"
    type="button"
  >
    <div
      class="c2"
    >
      <span
        class="c3 c4"
        color="txt-black-darkest"
        data-testid="bezier-react-select-trigger-text"
      >
        lorem ipsum
      </span>
      <div>
        가나다
      </div>
    </div>
    <svg
      class="c5"
      color="txt-black-darker"
      data-testid="bezier-react-icon"
      fill="none"
      foundation="[object Object]"
      height="16"
      marginbottom="0"
      marginleft="6"
      marginright="0"
      margintop="0"
      viewBox="0 0 24 24"
      width="16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clip-rule="evenodd"
        d="M19.707 8.793a1 1 0 0 1 0 1.414l-7 7a1 1 0 0 1-1.414 0l-7-7a1 1 0 0 1 1.414-1.414L12 15.086l6.293-6.293a1 1 0 0 1 1.414 0Z"
        fill="currentColor"
        fill-rule="evenodd"
      />
    </svg>
  </button>
</div>
`;
